pymc4/flow/executor.py

```diff
@@ -143,12 +143,17 @@ def __init__(
         self.deterministics = deterministics
         self.posterior_predictives = posterior_predictives
 
-    def collect_log_prob(self):
-        all_terms = itertools.chain(
+    def collect_log_prob_elemwise(self):
+        return itertools.chain(
             (dist.log_prob(self.all_values[name]) for name, dist in self.distributions.items()),
             (p.value for p in self.potentials),
         )
-        return sum(map(tf.reduce_sum, all_terms))
+
+    def collect_log_prob(self):
+        return sum(map(tf.reduce_sum, self.collect_log_prob_elemwise()))
+
+    def collect_unreduced_log_prob(self):
```

Comment:
this looks like a `ValueError` waiting to happen, unless all the elements are the same shape, right?

Reply:
Yes

We are doing automatic batching in all three big sampling functions: `sample_prior_predictive`, `sample` and `sample_posterior_predictive`. To be able to get the auto batching to work in some cases brings an over-head of extra `evaluate_model*` calls. Furthermore, the auto batching will have the overhead of always creating `SamplingState`s and other intermediate entities for every batch element. So in the end, I think that if someone wants to get their hands dirty to manually ensure that their model is well vectorized and batched, then they should have the means to do so.

I agree that this function is dangerous. It could be a `ValueError` waiting to happen, or a silently well broadcasting thing that leads to mysterious errors down the way, but it provides the necessary interface for a user to manually vectorize and batch their model.

Reply:
It's very hard to get the model to be well vectorized. If there are reductions, or tensor products, or even simple sums, the `batch_shape`s can lead to horribly wrong `log_prob` tensors. The automatic batching is super user friendly and awesome, but someone may ask for a more low level control of what goes on

Reply:
i'm fine leaving this -- was just wondering if I was right or wrong about it being dangerous :D